From 528b5af566ba9f30763b582b83983083e6dae720 Mon Sep 17 00:00:00 2001
From: Hauke Mehrtens <hauke.mehrtens@intel.com>
Date: Thu, 19 Nov 2020 18:27:34 +0100
Subject: [PATCH] PONRTSYS-8285: Misc improvements for the multicast filter
 bridge driver

This does some adaptions which should improve the usage inside of the
kernel.
* Add function dp_mcast_filter_active() instead of directly setting the
  variable. The function is now exported because this is set from a
  different kernel module.
* Adapt the comment style to the Linux networking documentation style
* Update the copyright
---
 .../mcast_bridge_filter/mcast_bridge_filter.c      | 48 ++++++++++------------
 include/net/mcast_bridge_filter.h                  | 12 ++----
 2 files changed, 25 insertions(+), 35 deletions(-)

diff --git a/drivers/net/datapath/mcast_bridge_filter/mcast_bridge_filter.c b/drivers/net/datapath/mcast_bridge_filter/mcast_bridge_filter.c
index 1cec7c571511..002578757b85 100644
--- a/drivers/net/datapath/mcast_bridge_filter/mcast_bridge_filter.c
+++ b/drivers/net/datapath/mcast_bridge_filter/mcast_bridge_filter.c
@@ -1,34 +1,34 @@
+// SPDX-License-Identifier: BSD-3-Clause OR GPL-2.0
 /******************************************************************************
  *
+ * Copyright (c) 2020 MaxLinear, Inc.
  * Copyright (c) 2018 Intel Corporation
  *
- * For licensing information, see the file 'LICENSE' in the root folder of
- * this software module.
- *
  ******************************************************************************/
-/*
- * This intercepts all packets transferred on any bridge in the system and
+/* This intercepts all packets transferred on any bridge in the system and
  * drops the IPv4 IGMP and the IPv6 MLD packets.
  */
 #include <linux/kernel.h>
-#include <linux/module.h>
 #include <linux/netfilter.h>
 #include <linux/netfilter_ipv4.h>
 #include <linux/netfilter_bridge.h>
 #include <linux/if_vlan.h>
 #include <linux/in.h>
-#include <linux/version.h>
-#include <net/ip.h>
-#include <net/ipv6.h>
+#include <linux/ip.h>
 #include <net/mcast_bridge_filter.h>
 
 /* max number of encapsulated vlan tags */
 #define MAX_VLAN_TAGS	3
 
-bool dp_mcast_vlan_activate = false;
+static bool dp_mcast_fliter_active;
 
-/*
- * Iterates over the VLAN tags and increases nhoff for each VLAN tag.
+void dp_mcast_filter_active(bool active)
+{
+	dp_mcast_fliter_active = active;
+}
+EXPORT_SYMBOL(dp_mcast_filter_active);
+
+/* Iterates over the VLAN tags and increases nhoff for each VLAN tag.
  * returns 0 on success and -EINVAL in case of an parsing error.
  */
 static int dp_mcast_vlan_iter(struct sk_buff *skb, __be16 *proto, int *nhoff)
@@ -51,8 +51,7 @@ static int dp_mcast_vlan_iter(struct sk_buff *skb, __be16 *proto, int *nhoff)
 	return 0;
 }
 
-/*
- * Returns 1 for an IPv6 MLD or MGM packet, 0, -EINVAL on an parsing error and
+/* Returns 1 for an IPv6 MLD or MGM packet, 0, -EINVAL on an parsing error and
  * 0 if it is not an IPv6 MLD or MGM packet.
  */
 static int dp_mcast_ipv6_is_mld_pkt(struct sk_buff *skb, int nhoff)
@@ -127,13 +126,13 @@ static unsigned int dp_mcast_br_hook_ipv6(struct sk_buff *skb, int nhoff)
 }
 
 static unsigned int dp_mcast_br_hook(void *priv, struct sk_buff *skb,
-				  const struct nf_hook_state *state)
+				     const struct nf_hook_state *state)
 {
 	__be16 proto = skb->protocol;
 	int nhoff = skb_network_offset(skb);
 	int err;
 
-	if (!dp_mcast_vlan_activate)
+	if (!dp_mcast_fliter_active)
 		return NF_ACCEPT;
 
 	if (eth_type_vlan(proto)) {
@@ -161,20 +160,17 @@ static struct nf_hook_ops dp_mcast_ops = {
 	.priority = NF_BR_PRI_BRNF,
 };
 
-int __init dp_mcast_init(void)
+static int __init dp_mcast_init(void)
 {
-#if (KERNEL_VERSION(4, 13, 0) > LINUX_VERSION_CODE)
 	return nf_register_hook(&dp_mcast_ops);
-#else
-	return nf_register_net_hook(&init_net, &dp_mcast_ops);
-#endif
 }
 
-void __exit dp_mcast_fini(void)
+static void __exit dp_mcast_fini(void)
 {
-#if (KERNEL_VERSION(4, 13, 0) > LINUX_VERSION_CODE)
 	nf_unregister_hook(&dp_mcast_ops);
-#else
-	nf_unregister_net_hook(&init_net, &dp_mcast_ops);
-#endif
 }
+module_init(dp_mcast_init);
+module_exit(dp_mcast_fini);
+
+MODULE_LICENSE("Dual BSD/GPL");
+MODULE_DESCRIPTION("PON multicast filter Driver");
diff --git a/include/net/mcast_bridge_filter.h b/include/net/mcast_bridge_filter.h
index 72952dc011e8..a0b4835e7246 100644
--- a/include/net/mcast_bridge_filter.h
+++ b/include/net/mcast_bridge_filter.h
@@ -1,21 +1,15 @@
+// SPDX-License-Identifier: BSD-3-Clause OR GPL-2.0
 /******************************************************************************
  *
+ * Copyright (c) 2020 MaxLinear, Inc.
  * Copyright (c) 2018 Intel Corporation
  *
- * For licensing information, see the file 'LICENSE' in the root folder of
- * this software module.
- *
  ******************************************************************************/
 
 #ifndef __PON_MCC_MCNF_H
 #define __PON_MCC_MCNF_H
 
-/* Init the IGMP / MLD bridge filter. */
-int __init dp_mcast_init(void);
-/* Uninit the IGMP / MLD bridge filter. */
-void __exit dp_mcast_fini(void);
-
 /* Activate the IGMP / MLD bridge filter. */
-extern bool dp_mcast_vlan_activate;
+void dp_mcast_filter_active(bool active);
 
 #endif /* __PON_MCC_MCNF_H */
